atoma-proxy/src/server/middleware.rs

```diff
@@ -587,20 +613,103 @@ pub async fn handle_locked_stack_middleware(
-        utils::lock_stack(&state.state_manager_sender, &mut req_parts, &endpoint)?;
-        req_parts
-            .headers
-            .insert(AUTHORIZATION, authorization_header);
-        let req = Request::from_parts(req_parts, Body::from(body_bytes));
-        if is_confidential_compute_endpoint(&endpoint) {
-            confidential_compute_middleware(state, req, next).await
-        } else {
+    match response.status() {
+        StatusCode::LOCKED => {
+            // Lock the current stack, in the Proxy's internal state
+            utils::lock_stack(&state.state_manager_sender, &mut req_parts, &endpoint)?;
+            req_parts
+                .headers
+                .insert(AUTHORIZATION, authorization_header);
+            let req = Request::from_parts(req_parts, Body::from(body_bytes));
+            if is_confidential_compute_endpoint(&endpoint) {
+                // NOTE: If this is a confidential compute request, the client already sent a request encrypted for a specific node.
+                // Acquiring a new stack in this case might simply select a new node, which means that the client will have to re-encrypt the request,
+                // with the new node's public key. For this reason, instead of being the proxy retrying to acquire a new stack, we let the client retry
+                // a couple of times, through Atoma's native SDKs.
+                return Err(AtomaProxyError::Locked {
+                    message: "Confidential compute requests should be retried".to_string(),
+                    endpoint: endpoint.to_string(),
+                });
+            }
+            // NOTE: For non-confidential requests, we can acquire a new stack directly from the proxy, and therefore we require a new authentication
+            // for the request.
             authenticate_middleware(state, req, next).await
         }
-    } else {
-        Ok(response)
+        StatusCode::TOO_EARLY => {
+            // NOTE: In this case, the node hasn't locked the stack immediately (has overestimated the compute units required for the request).
+            if is_confidential_compute_endpoint(&endpoint) {
+                // NOTE: If this is a confidential compute request, the client already sent a request encrypted for a specific node.
+                // Acquiring a new stack in this case might simply select a new node, which means that the client will have to re-encrypt the request,
+                // with the new node's public key. For this reason, instead of being the proxy retrying to acquire a new stack, we let the client retry
+                // a couple of times, through Atoma's native SDKs.
+                return Err(AtomaProxyError::UnavailableStack {
+                    message: "Confidential compute requests should be retried".to_string(),
+                    endpoint: endpoint.to_string(),
+                });
+            }
+            // We need to acquire a new stack for the request, to be able to retry
```

Comment:
Would we do a similar re-try when the resources are locked as well?

Reply:
In the case of a stack is locked, the proxy will try to acquire a new stack for you. In the confidential compute case it can't (because there is no guarantee the newly chosen node will be the same to which requests were encrypted in the first place).